module3/src/MyList.java

```diff
@@ -0,0 +1,697 @@
+     * @throws ClassCastException if the class of the specified element
+     *         prevents it from being added to this list
+     * @throws NullPointerException if the specified element is null and
+     *         this list does not permit null elements
+     * @throws IllegalArgumentException if some property of the specified
+     *         element prevents it from being added to this list
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *         (<tt>index &lt; 0 || index &gt; size()</tt>)
+     */
+    @Override
+    public void add(int index, Object element) {
+        throw new UnsupportedOperationException();//return;  //Stanislav said we do not need to implement it
+    }
+
+    /**
+     * Removes the element at the specified position in this list (optional
+     * operation).  Shifts any subsequent elements to the left (subtracts one
+     * from their indices).  Returns the element that was removed from the
+     * list.
+     *
+     * @param index the index of the element to be removed
+     * @return the element previously at the specified position
+     * @throws UnsupportedOperationException if the <tt>remove</tt> operation
+     *         is not supported by this list
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *         (<tt>index &lt; 0 || index &gt;= size()</tt>)
+     */
+    @Override
+    public T remove(int index) {
+        if (index < 0 || index >= size || size < 1) {
+            throw new IndexOutOfBoundsException(); //we can throw UnsupportedOperationException (or another exception) for case when size < 1
+        }
+        T toReturn = (T) elementData[index];
+        for (int i = index; i < size-1; i++) {
+            elementData[i] = elementData[i+1];
+        }
+        size--;
+        elementData[size] = null;
+        return toReturn;
+    }
+
+    /**
+     * Returns the index of the first occurrence of the specified element
+     * in this list, or -1 if this list does not contain the element.
+     * More formally, returns the lowest index <tt>i</tt> such that
+     * <tt>(o==null&nbsp;?&nbsp;get(i)==null&nbsp;:&nbsp;o.equals(get(i)))</tt>,
+     * or -1 if there is no such index.
+     *
+     * @param o element to search for
+     * @return the index of the first occurrence of the specified element in
+     *         this list, or -1 if this list does not contain the element
+     * @throws ClassCastException if the type of the specified element
+     *         is incompatible with this list
+     *         (<a href="Collection.html#optional-restrictions">optional</a>)
+     * @throws NullPointerException if the specified element is null and this
+     *         list does not permit null elements
+     *         (<a href="Collection.html#optional-restrictions">optional</a>)
+     */
+    @Override
+    public int indexOf(Object o) {
+        for(int i = 0; i < size; i++) {
+            if (o != null) {
```

Comment:
null check should be only one at the start of the method. You don't need to check it on each iteration, since `o` variable cannot be changed

Reply:
reworked